### PR TITLE
Collect `batch_size` from `training_args` and automatically pass it to dataloader

### LINKED ISSUE #392: Handling of `batch_size` is broken
In GitLab by @sharkovsky on Oct 12, 2022, 12:01

## update: bug description

I am changing this issue into a bug because I have discovered what I think is unexpected behaviour: when the researcher does not set a `batch_size` argument to their `training_data` function, the batch size in the training arguments is not correctly propagated to the `DataManager._loader_arguments`, and therefore not propagated to the data loader.

## additional confusing things

To my understanding, the way we currently handle `batch_size` seems a bit confusing. 

Here are some examples of what I mean:
1. Researchers are encouraged to declare `batch_size` as a parameter to their custom `training_data` function. In almost all our notebooks, we encourage them to define a default value, typically something like
```python
def training_data(self, batch_size = 48):
```
However, this default value will never be taken into account! It will always be superseded by the default value that we set in the `TrainingArgs` validator.

2. If the researcher forgets to declare `batch_size` as an argument to the `DataManager` returned by their `training_data` function, then whatever they set in the `training_args` will be silently ingored! 

I believe we should have a clearer way of handling this. 

My suggestion (but I'm open for discussion) would be:
- do not ask the researchers to manually put batch size anywhere. Even better, the `training_data` function should not take any arguments.
- automatically inject the batch size through some interface with the DataManager (could be a `set_batch_size` method, or a `append_to_loader_kwargs` method, etc..)
- the default value for `batch_size` in `TrainingArgs` should be 1, not 48, to avoid issues with smaller datasets.

### LINKED ISSUE #469: Use of batch_size argument, passed through training_args
In GitLab by @ibalelli on Jan 23, 2023, 10:06

``batch_size`` argument is given by the researcher as a ``training_args``, nevertheless in practice this is not used in the experiment: in all examples provided, the ``batch_size`` is further fixed as a default argument of the ``traning_data`` method when the training plan is customized by the researcher. This mean, that if one want to change the batch size, he finally has to modify ``training_data``rather than simply update the ``training_args``.

### PR DESCRIPTION
**Pull request by sharkovsky**
**PR description**

Closes #392 , closes #469 .

This PR introduces the following changes:
1. `batch_size` is **no longer** a valid key for `training_args`. Instead, it is replaced by `loader_args`, which is a generic dictionary of keyword arguments that will be passed to the data loader
2. the `training_data` function no longer takes `batch_size` as argument. Now, it doesn't take any arguments.
3. the `batch_size` is no longer manipulated by the researcher in the training plan. Instead, `batch_size` is **automatically injected** in the `DataManager`'s loader arguments, and ultimately passed on to the data loader.

Please read and review the updated documentation to find out more.

## Changes for the researcher

From a researcher point of view, this leads to the following changes.

### Definition of `training_data`
Before:
```python
class TrainingPlan:
    def training_data(self, batch_size = 1):
        dataset = MyDataset()
        loader_kwargs = {'batch_size': batch_size, 'shuffle': True}
        return DataManager(dataset, **loader_kwargs)
```
After:
```python
class TrainingPlan:
    def training_data(self):
        dataset = MyDataset()
        loader_kwargs = {'shuffle': True}
        return DataManager(dataset, **loader_kwargs)  # batch_size is collected from training_args and automatically passed to dataloader
```

### Changes to `training_args`
Before
```python
training_args = {
  'batch_size': 1
}
```
After
```python
training_args = {
  'loader_args': {
    'batch_size': 1,
    'drop_last: True
  }
}
```

### CODE REVIEW

**Comment by srcansiz** [comment]
Hi @sharkovsky, 

Thank you for the great effort of changing all the notebooks! I didn't check all one by one, I trust on that :) I commented only one issue that I found regarding the setting and getting loader arguments in `TrainingPlan`s. 

My another comment will be about unit tests, I saw that `codecov` complains about a few line that are not covered in your patch. https://github.com/fedbiomed/fedbiomed/pull/830/checks. 

Apart from these, I don't have anything to add. It can be merged  after the discussions are resolved.

File: fedbiomed/common/training_plans/_torchnn.py
Comment (sharkovsky) [reply fedbiomed/common/training_plans/_torchnn.py:227]: I agree that it should go into `BaseTrainingPlan`. I didn't do it for "symmetry", because the other similar methods are duplicated in the framework-specific plans. 
But I'll put it in the `BaseTrainingPlan`, and hopefully it will set the example for the other methods ;)

**Comment by sharkovsky** [comment]
@srcansiz I ended up moving all of the `*_arguments` methods to `BaseTrainingPlan`, I think it makes more sense this way and it reduces duplication.
 
A couple of comments:
- For `TorchTrainingPlan`, we need to override the `optimizer_args` getter because we call `update_optimizer_args`. 
- I also moved `aggregator_args` to `BaseTrainingPlan`. It seems to me that it should work fine, but they seem to be different from the other `*_args` (for instance, the researcher doesn't specify them).

**Comment by srcansiz** [comment]
Hi @sharkovsky, 

Thank you for the changes! Do you want me to merge it or wait for @ibalelli  and Lucia to review?

**Review by srcansiz — APPROVED** [review]

**Comment by manos-mark** [comment]
Although this is working great, I was not aware of those changes and I found the following bug.
I am still able to pass the `batch_size` argument using the old approach, and the sample size per epoch gets multiplied by the batch size, instead, it should be divided. 
 
This bug can be reproduced by:
a) Passing the `batch_size` argument using the old approach:
```
class TrainingPlan:
    def training_data(self, batch_size = 1):
        dataset = MyDataset()
        loader_kwargs = {'batch_size': batch_size, 'shuffle': True}
        return DataManager(dataset, **loader_kwargs)
```
        
b) Defining the `training_args` using the new approach:
```
training_args = {
  'loader_args': {
    'batch_size': 1,
    'drop_last: True
  }
}
```
I guess one possible solution, is to remove the `batch_size` argument from `training_data` function.

**Comment by mvesin** [comment]
Thank you so much @manos-mark !

We created #1207 to track this issue and will keep you updated soon !

**Pull request merged**
